NetKet/Graph/pygraph.hpp

```diff
@@ -39,46 +39,176 @@ namespace netket {
   .def("IsConnected", &AbstractGraph::IsConnected)		       \
   .def("Distances", &AbstractGraph::Distances);
 
-void AddGraphModule(py::module &m) {
+/// Given a Python iterable object constructs the edge list and (optionally)
+/// the colour map for the soon to be graph. `callback` is then called in one of
+/// the following ways:
+/// * `callback(edges, colour_map)` if the iterable contained elements of type
+/// `(int, int, int)`.
+/// * `callback(edges)` if the iterable contained elements of type `(int, int)`.
+///
+// TODO(twesterhout): We should probably split this into smaller functions...
+template <class Function>
+auto WithEdges(py::iterable xs, Function&& callback)
```

Comment:
We have to decide if we have simple (i.e. pybind11 free) constructors or not. Personally I am in favor of NOT having python-based constructors, see also comment to @femtobit . Can we make this using std objects instead?

Reply:
I see, you put it in `pygraph.hpp` to trick me :) 
Fine, as long as this dependence on pybind in contained in the py* files we can probably do this

Reply:
Constructors are pybind11-free, are they not? `WithEdges` is a _conversion_ function: it converts a Python iterable to a `std::vector<Edge>` while avoiding going through `std::vector<std::tuple<int, int>>`.

Reply:
> I see, you put it in `pygraph.hpp` to trick me :)
> Fine, as long as this dependence on pybind in contained in the py* files we can probably do this

Sorry, didn't notice this. Ignore my previous comment then :) Can I mark this as resolved then?

Reply:
Yes